src/main/java/com/wanderersoftherift/wotr/abilities/attachment/AbilityTracker.java

```diff
@@ -0,0 +1,123 @@
+import net.minecraft.core.Holder;
+import net.minecraft.nbt.Tag;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.Level;
+import net.neoforged.neoforge.attachment.IAttachmentHolder;
+import net.neoforged.neoforge.attachment.IAttachmentSerializer;
+import org.jetbrains.annotations.Nullable;
+
+public class AbilityTracker {
+
+    public static final IAttachmentSerializer<Tag, AbilityTracker> SERIALIZER = new AttachmentSerializerFromDataCodec<>(
+            Data.CODEC, AbilityTracker::new, AbilityTracker::data);
+
+    private final Multimap<Holder<TrackedAbilityTrigger.Type<?>>, TrackedAbility> abilities = ArrayListMultimap
+            .create();
+    private final IAttachmentHolder holder;
+
+    public AbilityTracker(IAttachmentHolder holder) {
+        this(holder, null);
+    }
+
+    public AbilityTracker(IAttachmentHolder holder, @Nullable Data data) {
+        this.holder = holder;
+    }
+
+    public static AbilityTracker forEntity(Entity entity) {
+        return entity.getData(WotrAttachments.ABILITY_TRACKER.get());
+    }
+
+    public static AbilityTracker forEntityNullable(Entity entity) {
+        var type = WotrAttachments.ABILITY_TRACKER.get();
+        if (!entity.hasData(type)) {
+            return null;
+        }
+        return entity.getData(type);
+    }
+
+    public static IAttachmentSerializer<Tag, AbilityTracker> getSerializer() {
+        return SERIALIZER;
+    }
+
+    private Data data() {
+        return null;
+    }
+
+    private Level getLevel() {
+        if (holder instanceof Entity entity) {
+            return entity.level();
+        }
+        return null;
+    }
+
+    public boolean triggerAbilities(TrackedAbilityTrigger activation) {
+        if (!(holder instanceof LivingEntity entity)) {
+            return false;
+        }
+        var result = false;
+        var typeHolder = getLevel().registryAccess()
+                .lookupOrThrow(WotrRegistries.Keys.TRACKED_ABILITY_TRIGGERS)
+                .wrapAsHolder(activation.type());
+        for (var tracked : abilities.get(typeHolder)) {
+            result |= holder.getData(WotrAttachments.ONGOING_ABILITIES)
+                    .activate(tracked.slot, tracked.slot.getContent(entity), tracked.ability);
+        }
+        return result;
+    }
+
+    public boolean hasAbilitiesOnTrigger(Holder<TrackedAbilityTrigger.Type<?>> activation) {
+        return !abilities.get(activation).isEmpty();
+    }
+
+    public void registerAbility(AbilityModifier abilityModifier, @Nullable WotrEquipmentSlot slot) {
+        registerAbility(abilityModifier.trigger(), abilityModifier.providedAbility(), slot);
+    }
+
+    public void registerAbility(
+            Holder<TrackedAbilityTrigger.Type<?>> trigger,
+            Holder<Ability> abilityHolder,
+            WotrEquipmentSlot slot) {
```

Comment:
Have you considered using ModifierSource here rather than slot? That should remove the need for the virtual slot and would be a more natural fit for the extensibility that ModiferSource offers.